### ISSUE TITLE
make `--time-unit months` more intuitive

### ISSUE THREAD

**Issue by richm** [elastic/curator#569]
I just spent several hours going through the code to find out why index selection based on months wasn't working.
For example, if today is April 15, and you want to delete indices that are 2 months older than today e.g. project.2016.02.14 (--prefix project.* --time-unit months --older-than 2 --timestring %Y.%m.%d), curator doesn't delete indices that are dated older than February 15, it deletes indices older than _February 1_.  That is, it goes back to the first day of the current month, _then_ goes back two whole months from that date.  One consequence is that if you have a cron job running curator every day, it won't delete anything for a whole month, then on the first day of the month, it will delete a months worth of indices.  Another consequence is that there is no way to delete indices that are exactly N months older than the current date - you can delete N*30 days ago, or N*4 weeks ago, but that is not the same as N months ago.
Since you can't very well fix `--time-unit months` without breaking existing usage of this counter-intuitive behavior, perhaps introduce another time unit e.g. `exact_months` or something like that.

**Comment by untergeek** [comment]
I'm sorry this is frustrating, but it is deliberate.  Were it not so, the Curator command line:

```
curator delete indices --timestring %Y.%m --time-unit months --older-than 1
```
would delete the January index on February 1.  Hence, Curator counts in _full_ months as a safety precaution against accidental deletion.  Of course, this causes confusion, and frustration for someone who wants to optimize the January index on February 1.  Because of this, you can use `--older-than 0` to get around this.

**Comment by untergeek** [comment]
> One consequence is that if you have a cron job running curator every day, it won't delete anything for a whole month, then on the first day of the month, it will delete a months worth of indices

This is incorrect. When using `--time-unit months`, each index is exactly 1 month, rather than counting days or weeks in a month.

Of course, this will become semi-moot for future releases of Curator (but only with Elasticsearch 2.0+) where index ages will be determined by the Stats API, rather than only by the index name pattern.

**Comment by PhaedrusTheGreek** [comment]
@untergeek 

I am hitting this issue as well.  Being that most people work in daily indices and curate monthly, this unexpected behaviour will probably be encountered by most curator users.    Perhaps we could improve [the documentation](https://www.elastic.co/guide/en/elasticsearch/client/curator/current/time-unit.html) around how to calculate monthly curating?

**Comment by richm** [comment]
We are recommending our users (of OpenShift aggregated logging) not to use `months`, but instead to use `--time-unit days --older-than $((Ndays * 30))` which should suffice.

**Comment by untergeek** [comment]
I'm sorry you're all having a less than ideal experience. This issue should be completely obviated in the upcoming 4.0 release of Curator, by way of the following changes:

Index timestamps can be calculated in 3 different ways:

1. Timestring in the index name
2. Index creation time 
3. Using the Elasticsearch stats_api to look at the youngest and oldest documents in an index, based on a timestamp field.

Rather than calculate to a month, day, or hour, Curator will use epoch time for these 3 methods, and will normalize epoch + milliseconds (or any other higher resolution) down to just epoch seconds.  The calculation of what is older than, or younger than a given epoch time will be by way of a reference point, also calculated as epoch time.  There will no longer be any concept of "whole months," which is what led to this issue.  If an index is older or younger than the reference point in time, it will be acted upon.

In order to make this work for users, more filtering options will almost certainly be necessary.  As a result, Curator 4.0 will no longer have any command-line options other than a version output, but will instead use a YAML configuration file.

I hope to have a beta version of 4.0 out in the next few weeks.  Curator 4.0 will _not_ be reverse compatible with any version of Elasticsearch older than 2.0.  It will also _not_ work with AWS ES, as Curator requires access to the cluster metadata, which endpoint is blocked there.

**Comment by sundarv85** [comment]
> Using the Elasticsearch stats_api to look at the youngest and oldest documents in an index, based on a timestamp field.

The stats_api method is only available from curator 4.0. Is that correct? Or will it work with curator 3.5.1 also.

> I hope to have a beta version of 4.0 out in the next few weeks

Do you already have a date in mind when the first curator 4.0 will be out.

**Comment by untergeek** [comment]
I hope to release 4.0 Beta before the end of the week.

**Comment by untergeek** [comment]
This should be addressed by #595 and #596 with Curator 4.0 imminent.

If you feel this is in error, please feel free to reopen this ticket or open another.
